Rework shard block and fraud proof (shard state transition) spec

hwwhww opened the pull request:
### Issue
1. Make the fraud-proofs.md spec in Python functions.
2. Add a basic test for shard transition.

### How did I fix it
1. `beacon-chain.md`
    1. Rework shard block
        1. Replace block wrapper with signable pattern
        2. Add `proposer_index` to shard block since the proposal index helps clients to validate the block in network level.
    2. Refactor `get_light_client_committee` a bit.
    3. Fix the wrong `get_shard_proposer_index` function calls
    4. Update `apply_shard_transition`
    5.  Fix typo `attestation.slot == state.slot` -> `attestation.data.slot == state.slot` in `is_winning_attestation`
    6. Check `verify_shard_transition_false_positives` **after** `process_operations`
    7. Fix `shard_attestations` filter in `process_crosslinks`: since attestations come from block, should use `attestation.data.slot + MIN_ATTESTATION_INCLUSION_DELAY == state.slot`
    8.  **[TBD]** Allow empty `light_client_signature` to make the tests pass
3. `phase1-fork.md`:
	* Add stub `PHASE_1_GENESIS_SLOT`, set it to `SLOTS_PER_EPOCH` now.
4. Rewrite the current fraud-proofs (shard state transition) spec in Python code
5. Add basic sanity test for shard transition and beacon block processing

terencechain commented on specs/phase1/beacon-chain.md:
```diff
@@ -674,7 +692,7 @@ def apply_shard_transition(state: BeaconState, shard: Shard, transition: ShardTr
-            headers.append(ShardSignableHeader(
+            headers.append(ShardBlockHeader(
                 shard_parent_root=shard_parent_root,
                 parent_hash=get_block_root_at_slot(state, get_previous_slot(state.slot)),
```
```suggestion
                beacon_parent_root =get_block_root_at_slot(state, get_previous_slot(state.slot)),
```

djrtwo commented on specs/phase1/fraud-proofs.md:
````diff
@@ -29,42 +32,264 @@ This document describes the shard transition function and fraud proofs as part o
+### Shard state transition function
+
+```python
+def shard_state_transition(beacon_state: BeaconState,
````
We're going to need to update gas in here

djrtwo commented on specs/phase1/fraud-proofs.md:
```diff
@@ -29,42 +32,264 @@ This document describes the shard transition function and fraud proofs as part o
+
+    if not is_empty_body(proposal.message):
+        # Apply state transition to shard_state.
+        shard_state_transition(
```
We might need to execute this transition because an empty block should have an affect on the gas limit (i believe).

djrtwo commented on specs/phase1/fraud-proofs.md:
```diff
@@ -29,42 +32,264 @@ This document describes the shard transition function and fraud proofs as part o
+
+### Make attestations
+
+Suppose you are a committee member on shard `shard` at slot `current_slot` and you have received shard blocks `shard_blocks`. Let `state` be the head beacon state you are building on, and let `QUARTER_PERIOD = SECONDS_PER_SLOT // 4`. `2 * QUARTER_PERIOD` seconds into slot `current_slot`, run `get_shard_transition(beacon_state, shard, shard_blocks)` to get `shard_transition`.
```
```suggestion
Suppose you are a committee member on shard `shard` at slot `current_slot` and you have received shard blocks `shard_blocks` since the latest successful crosslink for `shard` into the beacon chain. Let `state` be the head beacon state you are building on, and let `QUARTER_PERIOD = SECONDS_PER_SLOT // 4`. `2 * QUARTER_PERIOD` seconds into slot `current_slot`, run `get_shard_transition(beacon_state, shard, shard_blocks)` to get `shard_transition`.
```

hwwhww commented on specs/phase1/fraud-proofs.md:
```diff
@@ -29,42 +32,264 @@ This document describes the shard transition function and fraud proofs as part o
+        beacon_parent_root=hash_tree_root(beacon_parent_block),
+        signed_block=signed_block,
+    )
+    if shard_state.latest_block_root != transition.shard_states[offset_index].data:
```
Added `compute_shard_transition_data`! Agreed that `ShardState.data` should be clarified.
What do you think about `transition_digest`, to differentiate from `Root`?

djrtwo replied:
yeah, digest is a good call.

One thing to note is that we've generally avoided concatenation of bytes in a call to `hash` in favor of an explicit SSZ object and `HashTreeRoot` where ever possible. If we don't do it now, we'll likely need to soon.

protolambda commented on specs/phase1/beacon-chain.md:
```diff
@@ -761,7 +787,9 @@ def process_crosslink_for_shard(state: BeaconState,
         # Attestation <-> shard transition consistency
         assert shard_transition_root == hash_tree_root(shard_transition)
-        assert attestation.data.head_shard_root == shard_transition.shard_data_roots[-1]
+        assert attestation.data.head_shard_root == shard_transition.shard_data_roots[
```
The negative index is very pythonic IMHO. I like the python feature, but it makes it difficult to catch array index bugs, and other languages do not have the same.

djrtwo replied:
fine by me

protolambda commented on specs/phase1/beacon-chain.md:
```diff
@@ -772,11 +800,11 @@ def process_crosslink_for_shard(state: BeaconState,
         states_slots_lengths = zip(
             shard_transition.shard_states,
-            get_offset_slots(state, get_latest_slot_for_shard(state, shard)),
+            get_offset_slots(state, shard),
```
Should I prioritize mypy for remerkleable?

djrtwo replied:
I think getting it in there in the next couple of weeks will make dev of spec a little better

hwwhww commented:
Thanks @djrtwo!
All comments are applied or replied.

Changelog
1. Apply PR feedback
2. Rename & fix typos
3. `phase1-fork.md`:
	* Add stub `PHASE_1_GENESIS_SLOT`, set it to `SLOTS_PER_EPOCH` now.
4. `fraud-proofs.md` and `beacon-chain.md`
	1. https://github.com/ethereum/eth2.0-specs/pull/1703/commits/bb49c0891298ec5d88e540abf34855c0aac723a3 and https://github.com/ethereum/eth2.0-specs/pull/1703/commits/010201c65a5e45625f5fa4156e728662c1588329:
		1. Agreed that the skipped slot should affect  `gasprice`.
		2. Should have used `beacon_parent_root=get_block_root_at_slot(state, offset_slots[i])`, thanks!
                1. We *could* use the latest non-skipped slot block as the shard parent block, but it seems to not affect shard state at all? I simplified it to just `BeaconBlock(slot=offset_slot[i])` in https://github.com/ethereum/eth2.0-specs/pull/1703/commits/010201c65a5e45625f5fa4156e728662c1588329 now since only slot matters in shard transition now.
	2. https://github.com/ethereum/eth2.0-specs/pull/1703/commits/1aab7883ab9002c8dd60e0e40bae8f6fb9484162: Add `verify_shard_block_message`. Note that currently it won’t throw exception from `get_shard_transition` since `get_shard_transition` is a helper to select the valid proposals. Note that the honest validator guide may use `verify_shard_block_message`, so `verify_shard_block_message` and `is_valid_fraud_proof` would only be called upon receiving the shard blocks.

djrtwo commented on specs/phase1/beacon-chain.md:
```diff
@@ -727,7 +752,8 @@ def apply_shard_transition(state: BeaconState, shard: Shard, transition: ShardTr
 
     # Save updated state
-    state.shard_states[shard] = transition.shard_states[-1]
+    state.shard_states[shard] = transition.shard_states[len(transition.shard_states) - 1]
```
you're right. all good

djrtwo commented on specs/phase1/beacon-chain.md:
```diff
@@ -727,7 +752,8 @@ def apply_shard_transition(state: BeaconState, shard: Shard, transition: ShardTr
     # Save updated state
-    state.shard_states[shard] = transition.shard_states[-1]
+    state.shard_states[shard] = transition.shard_states[len(transition.shard_states) - 1]
+    assert state.slot > 0
```
Nice

I think we might still need `assert state.slot > PHASE_1_GENESIS_SLOT` because shard transitions from the prior slot do not yet exist. (or at least something handling this strange case)

djrtwo commented on specs/phase1/fraud-proofs.md:
```diff
@@ -29,42 +32,264 @@ This document describes the shard transition function and fraud proofs as part o
+            if validate_result:
+                assert verify_shard_block_signature(beacon_state, block)
+
+            shard_state_transition(
```
yeah, that seems like a good compromise for today!

djrtwo commented on tests/core/pyspec/eth2spec/test/helpers/shard_block.py:
```diff
@@ -0,0 +1,47 @@
+                      signed=False):
+    shard_state = beacon_state.shard_states[shard]
+    if slot is None:
+        slot = shard_state.slot
```
Should this be defaulting to `shard_state.slot + 1`? If the `beacon_state.shard_states[shard]` already has `slot` referenced, then the next block to be built should be the _next_ slot, not the same slot.

djrtwo replied:
Or maybe to `beacon_state.slot` as the normal case will be to built a shard for the same slot as the current beacon state

djrtwo commented on tests/core/pyspec/eth2spec/test/helpers/shard_block.py:
```diff
@@ -0,0 +1,47 @@
+
+    block = spec.ShardBlock(
+        shard_parent_root=shard_state.latest_block_root,
+        beacon_parent_root=spec.get_block_root_at_slot(beacon_state, spec.get_previous_slot(slot)),
```
In the normal case of building at the immediately next slot (the slot of the current beacon state), this call to `get_block_root_at_slot` will fail. In that case, it will instead need to be built against the `state.latest_block_header` and fill in the state root (see `build_empty_block` for an example)

djrtwo commented on tests/core/pyspec/eth2spec/test/phase_1/block_processing/test_process_crosslink.py:
```diff
@@ -0,0 +1,101 @@
+
+    committee_index = spec.CommitteeIndex(0)
+    shard = spec.compute_shard_from_committee_index(state, committee_index, state.slot)
+    body = b'1' * spec.MAX_SHARD_BLOCK_SIZE
```
I think it'd be easier to read if we use our common `b'\x'` notation (e.g. `b'\x42'`). This is clear that it is a full byte whereas the `b'1'` makes me a bit unsure the bitlength of "1" (without looking it up or testing it myself)

terencechain commented on specs/phase1/beacon-chain.md:
```diff
@@ -320,7 +328,7 @@ class ShardSignableHeader(Container):
     slot: Slot
     gasprice: Gwei
-    data: Bytes32
+    transition_digest: Bytes32
```
Should `transition_digest` eventually be `Root`?  
Looks like `compute_shard_transition_digest` will use ssz hash tree root for the whole digest

djrtwo replied:
Yeah, Root will be the appropriate type

terencechain commented on specs/phase1/fraud-proofs.md:
````diff
@@ -29,42 +32,259 @@ This document describes the shard transition function and fraud proofs as part o
+```
+
+```python
+def verify_shard_block_message(beacon_state: BeaconState,
````
might be best to move `verify_shard_block_message` and `verify_shard_block_signature` under helpers since they are only used by helper methods

terencechain commented on specs/phase1/fraud-proofs.md:
```diff
@@ -29,42 +32,259 @@ This document describes the shard transition function and fraud proofs as part o
+                         subkey: BLSPubkey,
+                         beacon_parent_block: BeaconBlock) -> bool:
+    # 1. Check if `custody_bits[offset_index][j] != generate_custody_bit(subkey, block_contents)` for any `j`.
+    shard = get_shard(beacon_state, attestation)
```
I think we can move `shard = get_shard(beacon_state, attestation)` to L136 to be closer to its usage

terencechain commented on specs/phase1/fraud-proofs.md:
````diff
@@ -29,42 +32,259 @@ This document describes the shard transition function and fraud proofs as part o
+### Shard state transition function
+
+```python
+def shard_state_transition(beacon_state: BeaconState,
````
The naming of `shard_state_transition` is pretty similar to `get_shard_transition`, but I can't think of a better name now...

hwwhww commented:
### Substantive changelog

1. Fix typo `attestation.slot == state.slot` -> `attestation.data.slot == state.slot` in `is_winning_attestation`
2. Check `verify_shard_transition_false_positives` **after** `process_operations`
3. Fix `shard_attestations` filter in `process_crosslinks`: since attestations come from block, should use `attestation.data.slot + MIN_ATTESTATION_INCLUSION_DELAY == state.slot`
4. **[TBD]** Allow empty `light_client_signature` to make the tests pass
5. Tests updates
    - Rework and refactor `test_process_crosslink`
    - Add basic phase 1 `test_blocks` ✨
        - `test_process_beacon_block_with_normal_shard_transition`: with basic on time shard attestation
        - `test_process_beacon_block_with_empty_proposal_transition`: The protocol acutally allows **on time** attestation that votes for "empty shard block proposal", i.e., creating shard transition with `shard_blocks=[]` for the skipped shard block.
    - After (3), some other tests failed. (e.g., `test_finality`), this PR updates the test helpers for backward compatibility, still kinda messy though.

terencechain commented on specs/phase1/beacon-chain.md:
```diff
@@ -694,27 +729,30 @@ def apply_shard_transition(state: BeaconState, shard: Shard, transition: ShardTr
-        block_length = transition.shard_block_lengths[i]
-        assert shard_state.gasprice == get_updated_gasprice(prev_gasprice, block_length)
+        # Verify correct calculation of gas prices and slots
+        assert shard_state.gasprice == compute_updated_gasprice(prev_gasprice, shard_block_length)
```
I think gas prices and slots assertion can be moved below `shard_state = transition.shard_states[i]`

Optional comment, that's what I did in Prysm : )

djrtwo commented:
> Fix shard_attestations filter in process_crosslinks: since attestations come from block, should use attestation.data.slot + MIN_ATTESTATION_INCLUSION_DELAY == state.slot

Note, `MIN_ATTESTATION_INCLUSION_DELAY` is essentially deprecated wth the new phase 1 design and _must_ be set to 1

djrtwo commented on tests/core/pyspec/eth2spec/test/phase_1/sanity/test_blocks.py:
```diff
@@ -0,0 +1,141 @@
+    assert shard_state.latest_block_root == shard_block.message.hash_tree_root()
+    assert shard_state != pre_shard_state
+    assert shard_state == shard_transition.shard_states[len(shard_transition.shard_states) - 1]
+
```
```suggestion
    assert shard_state.slot == state.slot - 1

```

djrtwo commented:
Looks good in general! I'm happy to merge soon and move on with subsequent testing PRs

terencechain commented on specs/phase1/beacon-chain.md:
```diff
@@ -761,7 +799,9 @@ def process_crosslink_for_shard(state: BeaconState,
         # Attestation <-> shard transition consistency
         assert shard_transition_root == hash_tree_root(shard_transition)
-        assert attestation.data.head_shard_root == shard_transition.shard_data_roots[-1]
+        assert attestation.data.head_shard_root == shard_transition.shard_data_roots[
```
I think `attestation` is out of scope here. It's outside of `for attestation in transition_attestations:`
Should it be `attestations[0]`? Assuming same shard transition root implies having the same shard head root

hwwhww replied:
@terencechain Nice catch!

I think it should be inside `for attestation in transition_attestations` here? Updated in https://github.com/ethereum/eth2.0-specs/pull/1703/commits/e758fb76c2ee67afa63a581c5b03022c6a739875

djrtwo replied:
looks good

hwwhww commented:
@djrtwo 

> Note, MIN_ATTESTATION_INCLUSION_DELAY is essentially deprecated wth the new phase 1 design and must be set to 1

👍 Added a comment to note that.

>  I'm happy to merge soon and move on with subsequent testing PRs

I refactored the tests and I'm happier with it now 😄
also squashed the reviewed commits by review cycles. I think it's okay to merge it now and then add more tests with other PRs.
